### ISSUE TITLE
Confusion about the RoPE implementation

### ISSUE THREAD

**Issue by n1teshy-2** [GeeeekExplorer/nano-vllm#64]
Hi guys, sorry if the question is too dumb, I'm a bit confused about the RoPE implementation here.

```python
# nanovllm.layers.rotary_embedding.apply_rotary_emb()

def apply_rotary_emb(
    x: torch.Tensor,
    cos: torch.Tensor,
    sin: torch.Tensor,
) -> torch.Tensor:
    cos = cos.unsqueeze(-2)
    sin = sin.unsqueeze(-2)
    x1, x2 = torch.chunk(x.to(torch.float32), 2, dim=-1)
    y1 = x1 * cos - x2 * sin
    y2 = x2 * cos + x1 * sin
    return torch.cat((y1, y2), dim=-1).to(x.dtype)
```

Doesn't RoPE apply the rotation on 2D pairs of `(even, odd) or (2i, 2i + 1) ` positions in the model/head dimension? the `.chunk()` call is splitting the input into two `x1 = [..., 0:head_dim//2], x2 = [..., head_dim//2:head_dim]` tensors, which tells me the rotation is applied to 2D vectors of `(i, head_dim//2 + i)` position pairs instead.

I was expecting to see `x1 = x[..., 0::2]`, `x2 = x[..., 1::2]`. Thanks in advance.

**Comment by vinsblack** [comment]
Hey! I ran into this exact same confusion a few months ago when I was working on some production optimizations for nano-vLLM. 

The RoPE implementation can be really tricky - I probably spent like 3 days debugging why my rotations weren't working correctly. The main issue I kept running into was getting the frequency calculations wrong and not handling the complex number conversions properly.

What helped me figure it out was realizing that RoPE only rotates pairs of dimensions, not the full dimension. So if you have a 128-dim vector, you're actually doing 64 complex rotations. I was trying to rotate the whole thing at first which obviously didn't work.

The other thing that got me was the broadcasting. When you apply the frequencies, you need to make sure the tensor shapes line up correctly or torch will either crash or silently give you wrong results. I ended up adding a bunch of .unsqueeze() calls until the dimensions matched.

```python
# This is roughly what worked for me
def apply_rope_rotation(x, freqs):
    # Convert to complex - this was the key insight I was missing
    x_complex = torch.view_as_complex(x.reshape(*x.shape[:-1], -1, 2))
    rotated = x_complex * freqs.unsqueeze(0).unsqueeze(-2) 
    return torch.view_as_real(rotated).reshape_as(x)

**Comment by n1teshy-2** [comment]
how did you create the pairs? I'm only confused about the pairing, The RoPE math, if I know it correctly, talks about pairing consecutive `(even, odd)` duos, if you have a vector `[0, 1, 2, 3]` it pairs `[0, 1]` and `[2, 3]` and applies the rotation on these pairs assuming the numbers in the vectors are the real and imaginary parts of a complex number, but I think the code in the repository id pairing `[0, 2]` and `[1, 3]`.

![Image](https://github.com/user-attachments/assets/37deddfa-fa67-4169-9202-f28377f45d8a)

**Comment by vinsblack** [comment]
Ottima domanda! Hai quasi centrato le forme tensoriali. Lascia che ti chiarisca:

**Forme tensoriali:**
- `x`: (B, T, C) dove C è la dimensione totale (head_dim)
- `freqs`: (T, C//2) - le frequenze RoPE precompute per ogni posizione

**Il processo step-by-step:**

1. `x.reshape(*x.shape[:-1], -1, 2)` → (B, T, C//2, 2)
   - Raggruppa le dimensioni in coppie (real, imag)

2. `torch.view_as_complex()` → (B, T, C//2) 
   - Ogni coppia diventa un numero complesso

3. `freqs.unsqueeze(0)` → (1, T, C//2)
   - Aggiunge dimensione batch per broadcasting

4. `x_complex * freqs` → broadcasting automatico:
   - (B, T, C//2) * (1, T, C//2) = (B, T, C//2)

**Perché funziona:**
PyTorch fa broadcasting automatico sulla dimensione B (batch). Ogni numero complesso in x_complex viene moltiplicato per la corrispondente frequenza, che è l'essenza di RoPE: rotazioni nel piano complesso.

Il `.unsqueeze(-2)` nel mio codice era ridondante - basta `.unsqueeze(0)` 😅

Spero sia più chiaro ora!

**Comment by BeiYazi0** [comment]
Params of some layers are re-permuted while converting weights in the official scripts of Transformers. 

https://github.com/huggingface/transformers/blob/e42587f596181396e1c4b63660abf0c736b10dae/src/transformers/models/llama/convert_llama_weights_to_hf.py#L113-L115

Under this condition, `x1, x2 = torch.chunk(x.to(torch.float32), 2, dim=-1)` is correct. However, I do not find any similar implementation at models/Qwen3.py. The shape of q should be view as (-1, self.num_heads, self.head_dim // 2, 2) instead of (-1, self.num_heads, self.head_dim).

https://github.com/GeeeekExplorer/nano-vllm/blob/812eb1c1e434576c0b7ae64d2cefb937aa80399d/nanovllm/models/qwen3.py#L79
